### PR TITLE
PWGHF: Add Ds task and selector 

### PR DESCRIPTION
**Pull request by stefanopolitano**
This PR contains the Ds -> phi pi -> K K pi task and selector. 
This PR continues the discussion from [AliceO2Group/O2Phyisics#96](https://github.com/AliceO2Group/O2Physics/pull/96) and[ AliceO2Group/Run3Analysisvalidarion#323](https://github.com/AliceO2Group/Run3Analysisvalidation/pull/323)

The last comments from the previous PR were addressed and the macros were moved to the proper folder (D2H).

### CODE REVIEW

**Review by fgrosa — APPROVED** [review]
The change in the event filtering code is trivial, so I approve it already

File: PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx
Comment (vkucera) [inline PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx:131]: You should set these based on the values of `topoDsToKKPi` and `topoDsToPiKK`.

File: PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx
Comment (vkucera) [inline PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx:159]: Is this part needed?

File: PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx
Comment (vkucera) [inline PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx:152]: Can a candidate be selected as DsKKPi and DsPiKK at the same time?

File: PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx
Comment (vkucera) [inline PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx:170]:
```suggestion
  return WorkflowSpec{adaptAnalysisTask<HfDsTokkpiCandidateSelector>(cfgc)};
```

File: PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx
Comment (vkucera) [inline PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx:29]:
```suggestion
struct HfDsToKKPiCandidateSelector {
```

File: PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx
Comment (vkucera) [inline PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx:35]: `requireTPC` is not used

File: PWGHF/D2H/Tasks/taskDs.cxx
Comment (vkucera) [inline PWGHF/D2H/Tasks/taskDs.cxx:29]: Move up

File: PWGHF/D2H/Tasks/taskDs.cxx
Comment (vkucera) [inline PWGHF/D2H/Tasks/taskDs.cxx:26]: `aod::hf_cand` should not be needed after merging https://github.com/AliceO2Group/O2Physics/pull/966

File: PWGHF/D2H/Tasks/taskDs.cxx
Comment (vkucera) [inline PWGHF/D2H/Tasks/taskDs.cxx:88]:
```suggestion
    registry.add("hPtVsYRecSigRecoPID", "3-prong candidates (RecoPID - matched);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
    registry.add("hPtVsYRecSigPromptRecoPID", "3-prong candidates (RecoPID - matched, prompt);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
    registry.add("hPtVsYRecSigNonPromptRecoPID", "3-prong candidates (RecoPID - matched, non-prompt);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
    registry.add("hPtVsYRecSigRecoTopol", "3-prong candidates (RecoTopol - matched);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
    registry.add("hPtVsYRecSigPromptRecoTopol", "3-prong candidates (RecoTopol - matched, prompt);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
    registry.add("hPtVsYRecSigNonPromptRecoTopol", "3-prong candidates (RecoTopol - matched, non-prompt);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
    registry.add("hPtVsYRecSigRecoSkim", "3-prong candidates (RecoSkim - matched);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
    registry.add("hPtVsYRecSigPromptRecoSkim", "3-prong candidates (RecoSkim - matched, prompt);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
    registry.add("hPtVsYRecSigNonPromptRecoSkim", "3-prong candidates (RecoSkim - matched, non-prompt);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
    registry.add("hPtVsYGen", "MC particles (matched);#it{p}_{T}^{gen.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
    registry.add("hPtVsYGenPrompt", "MC particles (matched, prompt);#it{p}_{T}^{gen.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
    registry.add("hPtVsYGenNonPrompt", "MC particles (matched, non-prompt);#it{p}_{T}^{gen.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
```

File: PWGHF/D2H/Tasks/taskDs.cxx
Comment (vkucera) [inline PWGHF/D2H/Tasks/taskDs.cxx:76]: What appears on the x axis title? `#it{p}_{T}^{gen.} (GeV/#it{c})` or `#it{p}_{T} (GeV/#it{c})`?

File: PWGHF/D2H/Tasks/taskDs.cxx
Comment (vkucera) [inline PWGHF/D2H/Tasks/taskDs.cxx:78]: You could define an axis for *y* to avoid repetition of `{100, -5., 5.}`.

File: PWGHF/D2H/Tasks/taskDs.cxx
Comment (vkucera) [inline PWGHF/D2H/Tasks/taskDs.cxx:132]: `tracks` are not used
```suggestion
  void processMC(soa::Join<aod::HfCandProng3, aod::HFSelDsToKKPiCandidate, aod::HfCandProng3MCRec> const& candidates,
                 soa::Join<aod::McParticles, aod::HfCandProng3MCGen> const& particlesMC, aod::BigTracksMC const&)
```

File: PWGHF/DataModel/HFCandidateSelectionTables.h
Comment (vkucera) [inline PWGHF/DataModel/HFCandidateSelectionTables.h:101]: Add documentation comments after `//!`

File: PWGHF/TableProducer/HFCandidateCreator3Prong.cxx
Comment (vkucera) [inline PWGHF/TableProducer/HFCandidateCreator3Prong.cxx:286]: Same comment as for `getMatchedMCRec`.

**Review by vkucera — CHANGES_REQUESTED** [review]

File: PWGHF/TableProducer/HFCandidateCreator3Prong.cxx
Comment (stefanopolitano) [reply PWGHF/TableProducer/HFCandidateCreator3Prong.cxx:217]: If I am not wrong, that should be fine because the decay vertex of the phi meson is not reconstructed since it does not travel much from the Ds decay vertex. In addition, I am going to implement a selection on the invariant mass of the kaon couple in order to ensure that the kaons come from a phi decay. Maybe also @fgrosa can comment on this.

Comment (vkucera) [reply PWGHF/TableProducer/HFCandidateCreator3Prong.cxx:217]: We are talking about MC matching here, not reconstruction. If the kaons come from a φ here, the Ds will be labelled as unmatched.

Comment (fgrosa) [reply PWGHF/TableProducer/HFCandidateCreator3Prong.cxx:217]: Hi @vkucera shouldn't the method look also for resonant decays? If not then it should be modified because we also have resonant decays in the Lc -> pKpi and D+ -> Kpipi for example, that should be considered when asking for the KKpi in the final state. 
For this specific case, I think that as long as the cut on the KK mass is not added at reco level (that I would suggest to add in a second step), all the Ds to KKpi should be considered in the MC (but keep in mind that in some MC productions we might have only Ds --> phi pi --> KKpi).

Comment (fgrosa) [reply PWGHF/TableProducer/HFCandidateCreator3Prong.cxx:217]: Ah indeed! It should be at least 2

Comment (fgrosa) [reply PWGHF/TableProducer/HFCandidateCreator3Prong.cxx:217]: Actually, there is the same problem for D+ and Xic+ in this file. @stefanopolitano could you please fix them as well? Thanks a lot!

File: PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx
Comment (stefanopolitano) [reply PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx:35]: Indeed, it is not needed inside the selector as the PID from TPC is already required in the trackselection step if I am not wrong (maybe @fgrosa can comment on this). However, the same line is present in the Dplus (https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/TableProducer/HFDplusToPiKPiCandidateSelector.cxx#L36) and Lc->K0sP (https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/TableProducer/HFLcK0sPCandidateSelector.cxx#L72) candidate selector and not used. Do you want me to remove this also from there?

Comment (fgrosa) [reply PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx:35]: Ciao @stefanopolitano, indeed I would remove it from here: in the `HFTrackIndexSkimsCreator` the quality selection on tracks is applied, so there is no need to add other requirements here (or at least nTPCcls > 0 is not a requirement we would need anyway). You can remove it I would say, thanks!

**Review by vkucera — CHANGES_REQUESTED** [review]

File: PWGHF/DataModel/HFCandidateSelectionTables.h
Comment (fgrosa) [reply PWGHF/DataModel/HFCandidateSelectionTables.h:99]: The exclamation mark in `ROOT` indicate that a data member is transient (see https://root.cern.ch/root/html534/guides/users-guide/InputOutput.html), so can it be the same here? If this is the case, it is important to keep them not only as a comment

Comment (vkucera) [reply PWGHF/DataModel/HFCandidateSelectionTables.h:99]: I checked and it seems only comments with `//!` go into the generated table/column documentation.

Comment (stefanopolitano) [reply PWGHF/DataModel/HFCandidateSelectionTables.h:99]: Ok, so I just leave the `\\!` as it was before, right?

Comment (vkucera) [reply PWGHF/DataModel/HFCandidateSelectionTables.h:99]: Yes

Comment (stefanopolitano) [reply PWGHF/DataModel/HFCandidateSelectionTables.h:99]: Ok, thanks!

File: PWGHF/D2H/Tasks/taskDs.cxx
Comment (stefanopolitano) [reply PWGHF/D2H/Tasks/taskDs.cxx:76]: I get `#it{p}_{T} (GeV/#it{c})`, I also checked  and it is in line with the other tasks.

File: PWGHF/TableProducer/HFTrackIndexSkimsCreator.cxx
Comment (vkucera) [inline PWGHF/TableProducer/HFTrackIndexSkimsCreator.cxx:885]: Do not rename these.

**Review by vkucera — CHANGES_REQUESTED** [review]

**Review by vkucera — DISMISSED** [review]
Thanks for addressing the comments. The current version looks fine to me.

**Comment by lvermunt** [comment]
Hi @stefanopolitano

I had a look at your Ds code. For most I agree with @vkucera that it is ready for merging besides the error highlighted by alibuild. However I have also a doubt about how you use the RecoDecay::getMatchedMCRec() function for the Ds. And also on your modification for the Dplus one.

Now I might be wrong since I just had a quick look (@fgrosa, maybe you can comment as well), but it seems to me you tag all resonant Ds -> X -> KKpi, and not just Ds -> phi pi -> KKpi as we did for the Run 2 analyses. Is this intended? If so, I would say you need to tag the different resonant channels for possibly different efficiency corrections (as currently already done for Lc -> pKpi).

For the Dplus, your modification of the depthMax to 2 is correct to also take the resonant K*(892)0 channel into account (as we did in Run 2). However, since the pdg code of the K* is not specified, I guess it also tags the D+ -> K+ K0s -> K+ pi pi channel. And this latter channel you don't want I would say.

Finally, your call to RecoDecay::getMother() in taskDs.cxx should also get the depthMax = 2 right?

Cheers, Luuk

**Review by lvermunt — COMMENTED** [review]
See https://github.com/AliceO2Group/O2Physics/pull/941#issuecomment-1222090701 for my review.

**Comment by vkucera** [comment]
Hi @lvermunt 

> Finally, your call to RecoDecay::getMother() in taskDs.cxx should also get the depthMax = 2 right?

`depthMax = -1` by default in `getMother()` so it does not make a difference for matched candidates.

**Comment by vkucera** [comment]
Hi @stefanopolitano , can you fix the error?

@lvermunt @xinyepeng Do you have any remaining comments?

**Comment by lvermunt** [comment]
Hi @vkucera, besides my two doubts on how the RecoDecay::getMatchedMCRec() function is used here which are not yet answered, and the error highlighted by alibuild, I don't have further comments.

Btw, if I understood correctly from a meeting earlier today, I think that Stefano is on holiday

**Comment by xinyepeng** [comment]
Hi @vkucera, I  agree with @lvermunt , nothing more from my side

**Comment by stefanopolitano** [comment]
Hi all, 
sorry for the late reply. I am addressing the points raised by @lvermunt. In particular:
- regarding the Ds I implemented a selection on the invariant mass f the KK couple (as was done in Run 2) to reject the candidates that do not include the phi decay. 
- concerning the Dplus I am not really sure how to implement this. Would it be better to look for the K0s PDG code for one of the intermediate particles of the decay chain in the HFCandidateCreator3Prong and reject all the candidates that present such code, or to implement this kind of selection in the HFDplusToPiKPiCandidateSelector?

**Comment by fgrosa** [comment]
> > Hi all, sorry for the late reply. I am addressing the points raised by @lvermunt. In particular:
> > ```
> > * regarding the Ds I implemented a selection on the invariant mass f the KK couple (as was done in Run 2) to reject the candidates that do not include the phi decay.
> > 
> > * concerning the Dplus I am not really sure how to implement this. Would it be better to look for the K0s PDG code for one of the intermediate particles of the decay chain in the HFCandidateCreator3Prong and reject all the candidates that present such code, or to implement this kind of selection in the HFDplusToPiKPiCandidateSelector?
> > ```
> 
> On the first point I agree that the current version of the code is consistent, though in a second PR we should include the selection on the massKK and consequently the selection of the phi resonance only. Regarding the second one I agree with @lvermunt, however this requires to modify the `RecoDecay` class to pass one or more "desired" intermediate states and I would also do it in a separated PR.
Sorry for the double comment: in the D+ case there should not be issues because the D+ -> K0SK+ would have different signs (K+pi-pi+ instead of K-pi+pi+), so there is no issue.

**Comment by lvermunt** [comment]
Hi @stefanopolitano, @fgrosa,

Thanks for your replies. 

I don't really follow your reply on the KK invariant mass. This is of course a good selection to include in the analysis. However, I was speaking about the tagging of the MC candidates with the getMatchedMCRec() function for the efficiency correction. This should always check the correct decay channel (or tag the various resonant ones), and not depend on a KK invariant mass selection being enabled. I think Fabrizio says the same. Anyhow, fine for me to address this in a 2nd PR.

Regarding the D+, you are of course right @fgrosa. Sorry for introducing noise there.

This PR is for me thus good to be merged.

Cheers, Luuk

**Review by lvermunt — APPROVED** [review]

**Comment by fgrosa** [comment]
Hi @lvermunt, 

> Hi @stefanopolitano, @fgrosa,
> 
> Thanks for your replies.
> 
> I don't really follow your reply on the KK invariant mass. This is of course a good selection to include in the analysis. However, I was speaking about the tagging of the MC candidates with the getMatchedMCRec() function for the efficiency correction. This should always check the correct decay channel (or tag the various resonant ones), and not depend on a KK invariant mass selection being enabled. I think Fabrizio says the same. Anyhow, fine for me to address this in a 2nd PR.
> 

The massKK selection selects the Ds decaying into phi, without that selection we are not measuring Ds -> phi pi -> KK pi, but Ds -> X -> KKpi, where X is any intermediate state. Hence, as long as we don't have that selection the code is consistent without requiring the phi (but I agree that the immediately next step is to add it).

> Regarding the D+, you are of course right @fgrosa. Sorry for introducing noise there.
> 
> This PR is for me thus good to be merged.
> 
> Cheers, Luuk

**Comment by lvermunt** [comment]
> Hi @lvermunt,
> 
> The massKK selection selects the Ds decaying into phi, without that selection we are not measuring Ds -> phi pi -> KK pi, but Ds -> X -> KKpi, where X is any intermediate state. Hence, as long as we don't have that selection the code is consistent without requiring the phi (but I agree that the immediately next step is to add it).

Sure, this I understand, and I agree the code is consistent enough (even though I'm also aware of your study for your PhD thesis that even with this massKK selection one still has a contamination of several percent by Ds -> f0(980)pi -> KKpi). My point was only that one shouldn't depend for the correct MC tagging on a selection that can be loosened or turned off. So that should be addressed in the future.

I approved and enabled the automatic merging. Unclear why the last pending check takes so long though

**Review by cterrevo — APPROVED** [review]

**Comment by vkucera** [comment]
@jgrosseo @ktf The errors in the macOS tests are unrelated. Can you please force the merge?

**Comment by jgrosseo** [comment]
You should be able to merge, macOS is not required.

**Comment by jgrosseo** [comment]
Maybe you miss the approval for `EventFiltering` ?

**Review by vkucera — APPROVED** [review]

**Comment by vkucera** [comment]
> Maybe you miss the approval for `EventFiltering` ?

Indeed

**Pull request merged**
